plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+  TSIOBufferReader output_reader;
+  TSVConn output_vc;
+  TSVIO output_vio;
+
+  TSAction pending_action;
+  TSVConn icap_vc;
+  TSVIO icap_vio;
+
+  std::string icap_header;
+  std::string http_header;
+  std::string chunk_length_str;
+  int64_t icap_reply_content_length;
+
+  int64_t http_body_chunk_length;
+  int64_t http_body_total_length_written;
+
+  bool eos_detected;
+
+  std::string err_msg;
+
+  TransformData(TSHttpTxn txnp);
+  ~TransformData();
+};
+
+/* Configurable paremeteres */
+static std::string server_ip;
+static int server_port;
+static int carp_port;
+static int debug_enabled;
+
+/* Stats for debug */
+static int scan_passed;
+static int scan_failed;
+static int icap_conn_failed;
+static int total_icap_invalid;
+static int icap_response_err;
+static int icap_write_failed;
+
+static int transform_handler(TSCont contp, TSEvent event, void *edata);
+static int transform_read_http_header_event(TSCont contp, TransformData *data, TSEvent event, void *edata);
+static int transform_send_error_msg(TSCont contp, TransformData *data);
+static int transform_bypass(TSCont contp, TransformData *data);
+static int transform_send_os_resp(TSCont contp, TransformData *data);
+
+TransformData::TransformData(TSHttpTxn txnp)
+{
+  state                          = STATE_BEGIN;
+  txn                            = txnp;
+  input_buf                      = NULL;
+  input_reader                   = NULL;
+  os_resp_buf                    = NULL;
+  os_resp_reader                 = NULL;
+  done_write                     = 0;
+  icap_resp_buf                  = NULL;
+  icap_resp_reader               = NULL;
+  output_buf                     = NULL;
+  output_reader                  = NULL;
+  output_vc                      = NULL;
+  output_vio                     = NULL;
+  pending_action                 = NULL;
+  icap_vc                        = NULL;
+  icap_vio                       = NULL;
+  http_body_chunk_length         = -1;
+  http_body_total_length_written = 0;
+  icap_header                    = std::string();
+  http_header                    = std::string();
+  chunk_length_str               = std::string();
+  icap_reply_content_length      = 0;
+  eos_detected                   = false;
+  err_msg                        = std::string();
+  server_reply_content_length    = 0;
+}
+
+TransformData::~TransformData()
+{
+  if (input_reader) {
+    TSIOBufferReaderFree(input_reader);
+  }
+  if (input_buf) {
+    TSIOBufferDestroy(input_buf);
+  }
+  if (os_resp_reader) {
+    TSIOBufferReaderFree(os_resp_reader);
+  }
+  if (os_resp_buf) {
+    TSIOBufferDestroy(os_resp_buf);
+  }
+  if (icap_resp_reader) {
+    TSIOBufferReaderFree(icap_resp_reader);
+  }
+  if (icap_resp_buf) {
+    TSIOBufferDestroy(icap_resp_buf);
+  }
+  if (output_reader) {
+    TSIOBufferReaderFree(output_reader);
+  }
+  if (output_buf) {
+    TSIOBufferDestroy(output_buf);
+  }
+  if (pending_action) {
+    TSActionCancel(pending_action);
+  }
+  if (icap_vc) {
```

Comment:
Maybe the VConnection should be aborted first before destroying the buffers/readers it interacts with.